### ISSUE TITLE
More concise display names for looped steps

### ISSUE THREAD

**Issue by james-simpson** [argoproj/argo-workflows#8937]
# Summary

I'd like to be able to customise the step name for a looped/fan out step generated by `withItems`/`withParam`. At the moment, the step name doesn't seem to be templatable - I get an error like 'must consist of lower case alphanumeric characters' when trying to submit the workflow in the UI.

E.g. could we do something like this, with `name: "{{item}}"`: 

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: loops-
spec:
  entrypoint: loop-example
  templates:
  - name: loop-example
    steps:
    - - name: "{{item}}"
        template: whalesay
        arguments:
          parameters:
          - name: message
            value: "{{item}}"
        withItems:
        - item-1
        - item-2
...

```

Maybe the best way to do it could be a separate templatable `displayName` field than overrides `name` in the UI if provided?

# Use Cases

We're using Argo Workflows to run end to end tests in parallel and it would be nice to customise the name of each fanned out step in the UI.

What we currently see:

![parallel-steps-1](https://user-images.githubusercontent.com/23004669/172807873-c8877780-e0e3-4e8d-af41-7d187122a7c1.png)

or

![parallel-steps-2](https://user-images.githubusercontent.com/23004669/172808520-cd897f72-0d33-407b-93cb-f66fa5ff18c5.png)

With custom/templated names we could see at a glance which scenario failed, and could easily find and click into a particular scenario we care about.

**Comment by CRASH-Tech** [comment]
Same issue

**Comment by andaag** [comment]
I got a similar issue - which I think is close enough it probably doesn't warrant its own issue.

I got very long parameters (a yaml blob) for some dynamic fan out jobs. For those the names become *extremely* long. Where I only really need 1 argument in the name/visible in the ui, and the rest visible in the "inputs/outputs" tab.

Considering the max length of these arguments is around 256kb, I think it's a bit much to stick all of them in the generated name 😀

**Comment by prakhar47b** [comment]
Same issue

**Comment by simonartige** [comment]
Same issue

**Comment by sfc-gh-rpeng** [comment]
Same issue

**Comment by agilgur5** [comment]
Potentially related: #2871

**Comment by JTarball** [comment]
same

**Comment by swzy** [comment]
Same underlying issue. Generated displayNames result in very long `workflow.failures` elements which in turn are causing unavoidable env variable size limitations.

**Comment by bhill74** [comment]
At the risk of sounding repetitive ... same issue for me.

**Comment by stefanondisponibile** [comment]
Same issue here.
Couldn't we use some annotation to use just on the UI? I believe Kubeflow uses `pipelines.kubeflow.org/task_display_name` in the same way. I don't think we need a custom _template_ name, we just need to customize how a pod name is displayed in the argo-server workflow ui. What do you think?

**Comment by KaiWaldrant** [comment]
Agree, we're looking for the same thing.

**Comment by Joibel** [comment]
This has been implemented since v3.7.0 by #14077, which added the `workflows.argoproj.io/display-name` template annotation. It is templated with the template's resolved inputs, so each fan-out child gets its own name, and the UI renders it in place of the generated `step(0:...)` name:

```yaml
  - name: cat-os-release
    annotations:
      workflows.argoproj.io/display-name: "os-{{inputs.parameters.image}}-{{inputs.parameters.tag}}"
    inputs:
      parameters:
      - name: image
      - name: tag
    ...
```

See the [loops walk-through](https://argo-workflows.readthedocs.io/en/latest/walk-through/loops/) and [annotations walk-through](https://argo-workflows.readthedocs.io/en/latest/walk-through/annotations/), plus `examples/loops-param-argument.yaml`. Note the value must match `^[a-zA-Z0-9][a-zA-Z0-9-.]{0,61}[a-zA-Z0-9]$`, and `{{item}}` is not available in the annotation directly — pass it through as an input parameter as the example does.

Closing as implemented. If your concern is specifically that the *default* generated name still embeds a large `withParam` item (rather than being able to override it), please open a new issue for that — it is a different change.
